### PR TITLE
Ping utility can optionally survive DNS failure

### PR DESCRIPTION
**Pull request by larham**
Previously, DNS was queried within the constructor for the `Ping` utility, so a failure in DNS always raised an exception. Now the DNS query takes place within the standard `run()` method. A DNS failure at that time will raise optionally, depending on the `validateAfter` parameter provided to the command.

### CODE REVIEW

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_ping.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        ping_command = Ping("testing", "doesNotExist.foo.com")
+        self.assertIn("-c 1 doesNotExist.foo.com", ping_command.cmdStr)
+
+    @patch('gppylib.commands.unix.socket.getaddrinfo', return_value=[(2, 2, 17, '', ('172.217.6.46', 0)), (2, 1, 6, '', ('172.217.6.46', 0)), (30, 2, 17, '', ('2607:f8b0:4005:809::200e', 0, 0, 0)), (30, 1, 6, '', ('2607:f8b0:4005:809::200e', 0, 0, 0))])
+    def test_happy_path_without_validation_pings(self, socket):
+        ping_command = Ping("testing", "google.com")
+        ping_command.run(validateAfter=False)
+        self.assertEquals(ping_command.get_results().rc, 0)
+
+    @patch('gppylib.commands.unix.socket.getaddrinfo', return_value=[(2, 2, 17, '', ('172.217.6.46', 0)), (2, 1, 6, '', ('172.217.6.46', 0)), (30, 2, 17, '', ('2607:f8b0:4005:809::200e', 0, 0, 0)), (30, 1, 6, '', ('2607:f8b0:4005:809::200e', 0, 0, 0))])
```
Comment (xinzweb) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_ping.py]: Can we combine the two happy path tests into one case, this will help reuse the single `@patch` command.

File: gpMgmt/bin/gppylib/commands/unix.py [hunk 2]
```diff
@@ -388,17 +388,29 @@ class Ping(Command):
...
         if curr_platform == LINUX or curr_platform == DARWIN or curr_platform == OPENBSD:
             # Get the family of the address we need to ping.  If it's AF_INET6
             # we must use ping6 to ping it.
-            addrinfo = socket.getaddrinfo(hostToPing, None)
-            if addrinfo and addrinfo[0] and addrinfo[0][0] == socket.AF_INET6:
-                pingToUse = SYSTEM.getPing6()
 
-        cmdStr = "%s -c 1 %s" % (pingToUse, hostToPing)
-        Command.__init__(self, name, cmdStr, ctxt, remoteHost)
+            try:
```
Comment (xinzweb) [inline gpMgmt/bin/gppylib/commands/unix.py]: Before the fix, we only run `socket.getaddrinfo` once to check the IPV6, now we are going to check in every single `run()` to check IPV6.

Is that possible to move the `try ... exception` in the `__init__()`, so that we only check the `IPV6` once?

The new logic in the `__init__()` would be just treated the error when finding the IPV6 failed, and still keep using IPV4 ping. In this case, it's still sound logic, since if the host is gone, it really doesn't matter we use IPV6 or IPV4.

This will minimize the change, and keep the logic closer to the original implementation.

We need to fix some of your tests accordingly. Thanks.

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_ping.py (hunk 1 above)
Comment (jchampio) [reply gpMgmt/bin/gppylib/operations/test/unit/test_unit_ping.py]: Unless we're using a framework that lets us parameterize our tests, I think having two separate assertions that can fail independently is worth the price of the duplication.

File: gpMgmt/bin/gppylib/commands/unix.py (hunk 2 above)
Comment (larham) [reply gpMgmt/bin/gppylib/commands/unix.py]: Trying to understand your perspective: do you interpret that the Ping() object is reused?  In all the production code I've seen, these python objects are instantiated, used once, and released for GC.  There is a lot of result state and stderr/stdout that is stored in the Command object itself, so if reuse was a goal, we'd probably see a method like `Command.reset()`, which does not exist.

Thus, actions that happen in constructor, vs. run() method, are both happening once in a single-use object.

What am I missing?

**Comment by jchampio** [comment]
Can you describe the desired use case? Mostly I'm wondering when it's okay for a ping to not actually ping.

**Comment by larham** [comment]
@jchampio This change is motivated by Kubernetes, which provides its own DNS service, wherein DNS entries are updated immediately when a container pops into or out of existence. It is important for Greenplum for Kubernetes to handle DNS "not found" results without considering any such failure a 100% fatal, stop-the-greenplum-cluster kind of failure.

In a traditional server farm, DNS is generally static, and Ping() is very likely to successfully get an IP address in its constructor, particularly if IP addresses are cached in /etc/hosts. 

The key usages of Ping() that are problematic are in the context of `GpInterfaceToHostNameCache.ping_hosts()` which is looking for failures to connect across all segments. For example, this is called during `gpstop`. Such usages are asking for *non-fatal* results (i.e. `Ping.run(validateAfter=False)`) so that they can simply report back any segment that cannot be reached. And vice-versa:  if Ping is supposed to raise an exception, it is called with `Ping.run(validateAfter=True)`.

In summary, the `Ping` command, and all subclasses of Command, are inherently supposed to use the `validateAfter` parameter to decide whether to raise or not.  Raising in the constructor is a bug that has not been evident in a server farm environment with static DNS.

**Comment by jchampio** [comment]
@larham Awesome, thanks for the detail!

File: gpMgmt/bin/gppylib/commands/unix.py (hunk 2 above)
Comment (xinzweb) [reply gpMgmt/bin/gppylib/commands/unix.py]: Thanks a lot @larham, yes that's context I am missing. I didn't realize the usage is just create the object once, call `run()` and throw it away. In that case, it doesn't matter where is the extra work, it's same amount of work overall.

As you pointed out in the code, I can see we can save a `super.run()` call on the `ping` command if we know the DNS failure in this fix. However on my side, I really want the constructor code (which is setting the `cmdStr` and `pingToUse` properties) to initialize its properties. The current design pattern of `Command` class hierarchy is basically let the super class `Command` to handle all the `run()` and let each subclass to add their own `__init__()` as well as any domain specific method.

In this case, I am proposing a slightly different fix, where we still keep the similar pattern as before (e.g. lever `super.run()` method as much as possible), but have the benefit of skipping `super.run()` if we know DNS failed.

```
    def __init__(self, name, hostToPing, ctxt=LOCAL, remoteHost=None, obj=None):
        self.hostToPing = hostToPing
        self.obj = obj
        self.pingToUse = findCmdInPath('ping')
        cmdStr = "%s -c 1 %s" % (self.pingToUse, self.hostToPing)

        # tried to see if we can use Ping6
        if curr_platform == LINUX or curr_platform == DARWIN or curr_platform == OPENBSD:
            # Get the family of the address we need to ping.  If it's AF_INET6
            # we must use ping6 to ping it.

            try:
                addrinfo = socket.getaddrinfo(self.hostToPing, None)
                if addrinfo and addrinfo[0] and addrinfo[0][0] == socket.AF_INET6:
                    self.pingToUse = SYSTEM.getPing6()
                    self.cmdStr = "%s -c 1 %s" % (self.pingToUse, self.hostToPing)
            except Exception as e:
                self.results = CommandResult(1, '', 'Failed to get ip address: ' + str(e), False, True)
                self.IsDNSFailure = true

        Command.__init__(self, name, cmdStr, ctxt, remoteHost)

    def run(self, validateAfter=False):
        if !self.isDNSFailure:
          super(Ping, self).run(validateAfter)
```

Comment (larham) [reply gpMgmt/bin/gppylib/commands/unix.py]: I like the advantage your proposed change demonstrates, where the `cmdStr` attribute is completely determined within the constructor.

However, there are disadvantages also: the constructor does network access, and any potential failure at construction time is temporally different than when `run()` is called.  In other words, the constructor is much "fatter", and if a user creates an object at time A and uses it at time B, the results of `socket.getaddrinfo` could be different at time B but we will have already recorded failure.

So I acknowledge the advantage of the suggested code, but I value "thinner" constructors and "fatter" `run()` methods that determine current ping status, regardless of when the object was created.

And I agree with your comment that we are both visitors to this code, and @nadeemg and team should weigh in on how they value the various tradeoffs presented here.

Comment (nadeemg) [reply gpMgmt/bin/gppylib/commands/unix.py]: I'm on the potato/po-tah-to fence. Mostly because we haven't been running in environments where things like IP addresses are likely to change. But imagining debugging those situtations, it seems like it might be valuable to get the ip address as close as possible to the time that we're going to use it. And there are lots of places in our code where we loop through creating and adding `Command` objects to a `WorkerPool`. So a delay between `init` and `run` seems likely. 

So, I have a preference for the original implementation, i.e. the thinner init and fatter run.
